src/measure_theory/covering/liminf_limsup.lean

```diff
@@ -0,0 +1,196 @@
+
+This file is a place to collect lemmas about liminf and limsup for subsets of a metric space
+carrying a doubling measure.
+
+## Main results:
+
+ * `blimsup_cthickening_mul_ae_eq`: the limsup of the closed thickening of a sequence of compact
+   subsets is unchanged if the sequence of distances is multiplied by a positive scale factor.
+
+-/
+
+open set filter metric measure_theory
+open_locale nnreal ennreal topological_space
+
+variables {α : Type*} [metric_space α] [proper_space α] [measurable_space α] [borel_space α]
+variables (μ : measure α) [is_locally_finite_measure μ] [is_doubling_measure μ]
+
+/-- This is really an auxiliary result en route to `blimsup_cthickening_mul_ae_eq`.
+
+NB: The `set : α` type ascription is present because of issue #16932 on GitHub. -/
+lemma blimsup_cthickening_ae_le_of_eventually_mul_le
+  (p : ℕ → Prop) {s : ℕ → set α} {M : ℝ} (hM : 0 < M)
+  {r₁ r₂ : ℕ → ℝ} (hr : tendsto r₁ at_top (𝓝[>] 0)) (hMr : ∀ᶠ i in at_top, M * r₁ i ≤ r₂ i) :
+  (blimsup (λ i, cthickening (r₁ i) (s i)) at_top p : set α) ≤ᵐ[μ]
+  (blimsup (λ i, cthickening (r₂ i) (s i)) at_top p : set α) :=
+begin
+  /- Sketch of proof:
+
+  Assume that `p` is identically true for simplicity. We may also assume that `M < 1` and `0 ≤ r₁`.
+  Let `Y₁ i = cthickening (r₁ i) (s i)`, define `Y₂` similarly except using `r₂`, and let
+  `(Z i) = ⋃_{j ≥ i} (Y₂ j)`. Our goal is equivalent to showing that `μ ((limsup Y₁) \ (Z i)) = 0`
+  for all `i`.
+
+  Assume for contradiction that `μ ((limsup Y₁) \ (Z i)) ≠ 0` for some `i` and let
+  `W = (limsup Y₁) \ (Z i)`. Apply Lebesgue's density theorem to obtain a point `d` in `W` of
+  positive density. Since `d ∈ limsup Y₁`, there is a subsequence of `j ↦ Y₁ j`, indexed by
+  `f 0 < f 1 < ...`, such that `d ∈ Y₁ (f j)` for all `j`. For each `j`, we may thus choose
+  `w j ∈ s (f j)` such that `d ∈ B j`, where `B j = closed_ball (w j) (r₁ (f j))`. Note that
+  since `d` has positive density `μ (W ∩ (B j)) / μ (B j) → 1`.
+
+  We obtain our contradiction by showing that there exists `η < 1` such that
+  `μ (W ∩ (B j)) / μ (B j) ≤ η` for sufficiently large `j`. In fact we claim that `η = 1 - C⁻¹`
+  is such a value where `C` is the scaling constant of `M⁻¹` for the doubling measure `μ`.
+
+  To prove the claim, let `b j = closed_ball (w j) (M * r₁ (f j))` and for given `j` consider the
+  sets `b j` and `W ∩ (B j)`. These are both subsets of `B j` and are disjoint for large enough `j`
+  since `M * r₁ j ≤ r₂ j` and thus `b j ⊆ Z i ⊆ Wᶜ`. We thus have:
+  `μ (b j) + μ (W ∩ (B j)) ≤ μ (B j)`. Combining this with `μ (B j) ≤ C * μ (b j)` we obtain
+  the required inequality. -/
+  suffices : ∀ {s : ℕ → set α} (hs : ∀ i, is_closed (s i))
+    {r₁ r₂ : ℕ → ℝ} (hr : tendsto r₁ at_top (𝓝[>] 0)) (hrp : 0 ≤ r₁)
+    {M : ℝ} (hM : 0 < M) (hM' : M < 1) (hMr : ∀ᶠ i in at_top, M * r₁ i ≤ r₂ i),
+    (blimsup (λ i, cthickening (r₁ i) (s i)) at_top p : set α) ≤ᵐ[μ]
+    (blimsup (λ i, cthickening (r₂ i) (s i)) at_top p : set α),
```

Comment:
I'd separate out the lemma in two lemmas, the first one assuming that the sets are closed, and the second one doing the reduction to closed sets, just to make things easier to read.

Reply:
OK I've broken out a lemma which reduces to closed sets (as well as to `M < 1` and `0 ≤ r₁`).